### ISSUE TITLE
[rllib] How to modify trainbatch

### ISSUE THREAD

**Issue by alexanderimanicowenrivers** [ray-project/ray#8266]
hey,

1) I am using the python version, and specifically using the torch ppo policy. I am trying to adapt Rllib so it can handle constrained reinforcement learning environments, which have constraint information all put into the gym info dict by default. What I would love is to be able to do train_batch[SampleBach.Cost] as we do with train_batch[SampleBach.Rewards] so I can use this information in my PPO loss. 

Effectively, you can access this cost in the gym environment, when taking a step we get obs, r, info, done = env.step(a), you can take info['cost'] to recover this value. Note that cost will be exactly the same shapes, type everything as a reward. So if there is a way to easily add in another reward into the train_batch that will also work? 

2), how may I do a full batch sample instead of a single batch sample? 

3) I want a training loop where before I commence my PPO mini-batch updates I do a global update on a constraint parameter just once, what file would I need to modify to do this, as I can only see access to the train_batch file, although if we do 80 sgd updates then anything inside this file will be updated 80 times, whereas we want it to be outside this loop and thus updated only once.

Thanks in advance!

Alex

**Comment by ericl** [comment]
1) With torch this is particularly easy, the info dict is already available in `train_batch["infos"]` in `def ppo_surrogate_loss(policy, model, dist_class, train_batch)`. You can customize the policy with a loss that includes the costs from the info dicts (after converting to a torch tensor).

2) Easiest way is probably to edit the execution plan here: https://github.com/ray-project/ray/blob/master/rllib/agents/ppo/ppo.py#L218 (enabled in latest version of RLlib)

add before TrainOneStep, add a `.for_each(update_constraints)` where your update_constraints function can access the full workers / policy state.

There's probably other easier places to compute it, for example in trajectory postprocessing, not sure the requirements you have.

**Comment by bentzinir** [comment]
Hi there,
Is there a way to access `train_batch["infos"]` in TensorFlow mode?
Thanks!
Nir
